File: ros_ws/src/navigation_collisions/src/collisions_subscriptions.py
import math, json
import rospy
import tf2_ros, tf

from obstacles_stack import Map, ObstaclesStack
from collisions_robot import Robot
from collisions_engine import Point, Position, Velocity, SegmentObstacle, RectObstacle, CircleObstacle

from memory_map.srv import MapGet
from navigation_navigator.msg import Status
from drivers_ard_asserv.msg import RobotSpeed
from recognition_objects_classifier.msg import ClassifiedObjects
from ai_game_manager import StatusServices


class CollisionsSubscriptions(object):
    def __init__(self):
        # Preparing to get the robot's position, belt frame_id transform.
        self._tf2_pos_buffer = tf2_ros.Buffer(cache_time=rospy.Duration(5.0))
        self._tf2_pos_listener = tf2_ros.TransformListener(self._tf2_pos_buffer)
        self._transform_listener = tf.TransformListener()

        # Callback buffers
        self._nav_status = 0 #STATUS_IDLE
        self._robot_path_waypoints = []
        self._vel_linear  = 0.0
        self._vel_angular = 0.0

        # Subscribing to dependencies
        rospy.Subscriber("/navigation/navigator/status", Status, self._on_nav_status)
        rospy.Subscriber("/recognition/objects_classifier/objects", ClassifiedObjects, self._on_classifier)
        rospy.Subscriber("/drivers/ard_asserv/speed", RobotSpeed, self.on_robot_speed)

        self.game_status = StatusServices("navigation", "collisions", None, self._on_game_status)

    def send_init(self, success = True):
        self.game_status.ready(success)

    def create_robot(self):
        try: # Getting the robot shape and creating the robot instance
            robot_type = rospy.get_param("/robot").lower()
            map_get_client = rospy.ServiceProxy("/memory/map/get", MapGet)
            map_get_client.wait_for_service(2.0)
            shape = json.loads(map_get_client("/entities/" + robot_type + "/shape/*").response)
            if not shape["type"] == "rect":
                raise ValueError("Robot shape type not supported here.")
        except Exception as e:
            rospy.logerr("ERROR Collisions couldn't get the robot's shape from map : " + str(e))
            shape = {"width": 0.4, "height": 0.25}
        return Robot(shape["width"], shape["height"]) # Can create a rect or circle

    def update_robot(self):
        new_pos = self._update_robot_pos()
        if new_pos is not None:
            Map.Robot.update_position(new_pos)

        if self._nav_status is not None:
            Map.Robot.update_status(self._nav_status)
        if self._robot_path_waypoints is not None and len(self._robot_path_waypoints) > 0:
            Map.Robot.update_waypoints(self._robot_path_waypoints)

        Map.Robot.update_velocity(self._vel_linear, self._vel_angular)

    def _update_robot_pos(self):
        try:
            t = self._tf2_pos_buffer.lookup_transform("map", "robot", rospy.Time())
            tx, ty = t.transform.translation.x, t.transform.translation.y
            rz = self._quaternion_to_euler_angle(t.transform.rotation)[2]
            return (tx, ty, rz)
        except Exception as e:
            return None

    def _on_game_status(self, msg):
        pass

    def _on_nav_status(self, msg):
        self._nav_status = msg.status
        self._robot_path_waypoints = [Point(point.x, point.y) for point in msg.currentPath]

    def _on_classifier(self, msg):
        new_belt = []
        for rect in msg.unknown_rects:
            if rect.header.frame_id not in ["map", "/map"]:
                rospy.logwarn("Belt rect not in /map tf frame, skipping.")
                continue

            new_belt.append(RectObstacle(Position(rect.x, rect.y,\
                                                  rect.a),\
                                                  rect.w, rect.h))

        if len(new_belt) > 0:
            ObstaclesStack.updateBeltPoints(new_belt)


        new_lidar = []

        for segment in msg.unknown_segments:
            if segment.header.frame_id not in ["map", "/map"]:
                rospy.logwarn("Lidar segment not in /map tf frame, skipping.")
                continue

            new_lidar.append(SegmentObstacle(Position(segment.segment.first_point.x, segment.segment.first_point.y),
                                             Position(segment.segment.last_point.x,  segment.segment.last_point.y)))
        for circle in msg.unknown_circles:
            if circle.header.frame_id not in ["map", "/map"]:
                rospy.logwarn("Lidar circle not in /map tf frame, skipping.")
                continue
            vel_d = math.sqrt(circle.circle.velocity.y ** 2 + circle.circle.velocity.x ** 2)
            vel_a = math.atan2(circle.circle.velocity.y, circle.circle.velocity.x)
            new_lidar.append(CircleObstacle(Position(circle.circle.center.x, circle.circle.center.y, angle = vel_a),
                                            circle.circle.radius, velocity = Velocity(circle.circle.radius * 2, circle.circle.radius * math.sqrt(3.0) / 2.0,
                                            vel_d, 0.0)))

        if len(new_lidar) > 0:
            ObstaclesStack.updateLidarObjects(new_lidar)

    def on_robot_speed(self, msg):
        self._vel_linear = msg.linear_speed
        self._vel_angular = 0.0

    def _quaternion_to_euler_angle(self, quaternion):
        # https://en.wikipedia.org/wiki/Conversion_between_quaternions_and_Euler_angles
        w, x, y, z = quaternion.w, quaternion.x, quaternion.y, quaternion.z
        t0 = +2.0 * (w * x + y * z)
        t1 = +1.0 - 2.0 * (x * x + y ** 2)
        X = math.degrees(math.atan2(t0, t1))

        t2 = +2.0 * (w * y - z * x)
        t2 = +1.0 if t2 > +1.0 else t2
        t2 = -1.0 if t2 < -1.0 else t2
        Y = math.degrees(math.asin(t2))

        t3 = +2.0 * (w * z + x * y)
        t4 = +1.0 - 2.0 * (y ** 2 + z * z)
        Z = math.atan2(t3, t4)
        return (X, Y, Z)

File: ros_ws/src/navigation_collisions/src/collisions/collisions_subscriptions.cpp
#include "collisions/collisions_subscriptions.h"

#include "collisions/shapes/circle.h"
#include "collisions/shapes/rectangle.h"
#include "collisions/shapes/segment.h"
#include "collisions/obstacle_velocity.h"

#include "memory_map/MapGet.h"

#include <nlohmann/json.hpp>
#include <ros/duration.h>

#include <cmath>
#include <string>

using namespace nlohmann;

const double CACHE_TIME_TF2_BUFFER = 5.0;
const std::size_t SIZE_MAX_QUEUE = 10;

const std::string NAVIGATOR_STATUS_TOPIC = "navigation/navigator/status";
const std::string OBJECTS_TOPIC          = "recognition/objects_classifier/objects";
const std::string ASSERV_SPEED_TOPIC     = "drivers/ard_asserv/speed";
const std::string MAP_GET_SERVER         = "memory/map/get";

const std::string NODE_NAME      = "collisions";
const std::string NAMESPACE_NAME = "navigation";

const std::string MAP_TF_FRAME          = "map";
const std::string ROBOT_TF_FRAME        = "robot";
const std::string PARAM_ROBOT_TYPE      = "/robot";
const std::string DEFAULT_ROBOT_NAME    = "gr";
const double      DEFAULT_ROBOT_WIDTH   = 0.4;
const double      DEFAULT_ROBOT_HEIGHT  = 0.25;

#include <iostream>

Position quaternionToEuler(geometry_msgs::Quaternion quaternion);

inline constexpr double radToDegrees(double angle) {
    return angle * M_PI / 180;
}

CollisionsSubscriptions::CollisionsSubscriptions(ros::NodeHandle& nhandle):
    obstaclesStack_(std::make_shared<ObstaclesStack>()),
    tf2PosBuffer_(ros::Duration(CACHE_TIME_TF2_BUFFER)),
    tf2PosListener_(tf2PosBuffer_)
{
    std::cerr << "Subscribing to " << ASSERV_SPEED_TOPIC << std::endl;
    asservSpeedSubscriber_ = nhandle.subscribe(
        ASSERV_SPEED_TOPIC,
        SIZE_MAX_QUEUE,
        &CollisionsSubscriptions::onAsservSpeed,
        this
    );
    navigatorStatusSubscriber_ = nhandle.subscribe(
        NAVIGATOR_STATUS_TOPIC,
        SIZE_MAX_QUEUE,
        &CollisionsSubscriptions::onNavStatus,
        this
    );
    
    objectsSubscriber_ = nhandle.subscribe(
        OBJECTS_TOPIC,
        SIZE_MAX_QUEUE,
        &CollisionsSubscriptions::onObjects,
        this
    );
    
    gameStatus_ = std::make_unique<StatusServices>(NAMESPACE_NAME, NODE_NAME, nullptr, [this](const ai_game_manager::GameStatus::ConstPtr& status) {
        this->onGameStatus(status);
    });
}

void CollisionsSubscriptions::sendInit(bool success) {
    gameStatus_->setReady(success);
}

CollisionsSubscriptions::RobotPtr CollisionsSubscriptions::createRobot(ros::NodeHandle& nhandle)
{
    std::lock_guard<std::mutex> lock(mutexRobot_);
    double width, height;
    std::string robotName = fetchRobotName(nhandle);
    try {
        auto mapGetClient = nhandle.serviceClient<memory_map::MapGet>(MAP_GET_SERVER);
        memory_map::MapGet msg;
        msg.request.request_path = "/entities/" + robotName + "/shape/*";
        ROS_INFO_STREAM("Waiting for service \"" << MAP_GET_SERVER << "\"");
        mapGetClient.waitForExistence();
        ROS_INFO_STREAM("Service found or timed out");
        if (!mapGetClient.call(msg) || !msg.response.success)
            throw ros::Exception("Call failed.");
        json shape = json::parse(msg.response.response);
        if (shape["type"] != "rect")
            throw ros::Exception("Shape '" + shape.at("type").get<std::string>() + "' not allowed.");
        width = shape["width"];
        height = shape["height"];
    } catch(const ros::Exception& e) {
        ROS_WARN_STREAM("Error when trying to contact '" + MAP_GET_SERVER + "' : " + e.what() + " Falling back to default robot's size value.");
        width = DEFAULT_ROBOT_WIDTH;
        height = DEFAULT_ROBOT_HEIGHT;
    } catch(...) {
        ROS_WARN_STREAM("Unknown error when trying to contact '" + MAP_GET_SERVER + "'. Falling back to default robot's size value..");
        width = DEFAULT_ROBOT_WIDTH;
        height = DEFAULT_ROBOT_HEIGHT;
    }
    robot_ = std::make_shared<Robot>(width, height);
    return robot_;
}

void CollisionsSubscriptions::updateRobot()
{
    std::lock_guard<std::mutex> lock(mutexRobot_);
    if (!robot_)
        return;
    auto newPos = updateRobotPos();
    if (newPos != Position(0,0,0)) {
        robot_->setPos(newPos);
    }
    
    robot_->updateStatus(navStatus_);
    
    if (!robotPathWaypoints_.empty()) {
        robot_->updateWaypoints(robotPathWaypoints_);
    }
    robot_->updateVelocity(velLinear_, velAngular_);
}


void CollisionsSubscriptions::onAsservSpeed(const drivers_ard_asserv::RobotSpeed::ConstPtr& speed)
{
    std::lock_guard<std::mutex> lock(mutexRobot_);
    velLinear_ = speed->linear_speed;
    velAngular_ = 0.0;
}


void CollisionsSubscriptions::onGameStatus(const ai_game_manager::GameStatus::ConstPtr& status) {
    // Nothing ?
}

void CollisionsSubscriptions::onNavStatus(const navigation_navigator::Status::ConstPtr& status)
{
    std::lock_guard<std::mutex> lock(mutexRobot_); // Needed ?
    if (status->status == status->NAV_IDLE)
        navStatus_ = Robot::NavStatus::IDLE;
    else if (status->status == status->NAV_NAVIGATING)
        navStatus_ = Robot::NavStatus::NAVIGATING;
    
    robotPathWaypoints_.clear();
    for (auto point: status->currentPath)
        robotPathWaypoints_.emplace_back(point.x, point.y, point.theta);
}

void CollisionsSubscriptions::onObjects(const recognition_objects_classifier::ClassifiedObjects::ConstPtr& objects)
{
    std::vector<std::shared_ptr<Obstacle>> newBelt, newLidar;
    
    for (auto rect: objects->unknown_rects) {
        if (rect.header.frame_id != MAP_TF_FRAME && rect.header.frame_id != "/" + MAP_TF_FRAME) {
            ROS_WARN_STREAM("Belt rect not in /" << MAP_TF_FRAME << " tf frame, skipping.");
            continue;
        }
        auto rectShape = std::make_shared<CollisionsShapes::Rectangle>(
            Position(rect.x, rect.y, rect.a),
            rect.w,
            rect.h
        );
        newBelt.emplace_back(std::make_shared<Obstacle>(rectShape));
    }
    if (!newBelt.empty()) {
        obstaclesStack_->updateBeltPoints(newBelt);
    }
    
    for (auto seg: objects->unknown_segments) {
        if (seg.header.frame_id != MAP_TF_FRAME && seg.header.frame_id != "/" + MAP_TF_FRAME) {
            ROS_WARN_STREAM("Lidar segment not in /" << MAP_TF_FRAME << " tf frame, skipping.");
            continue;
        }
        auto segShape = std::make_shared<CollisionsShapes::Segment>(
            Point(seg.segment.first_point.x, seg.segment.first_point.y),
            Point(seg.segment.last_point.x, seg.segment.last_point.y)
        );
        newLidar.emplace_back(std::make_shared<Obstacle>(segShape));
    }
    for (auto circ: objects->unknown_circles) {
        if (circ.header.frame_id != MAP_TF_FRAME && circ.header.frame_id != "/" + MAP_TF_FRAME) {
            ROS_WARN_STREAM("Lidar circle not in /" << MAP_TF_FRAME << " tf frame, skipping.");
            continue;
        }
        double velDist = std::hypot(circ.circle.velocity.x, circ.circle.velocity.y);
        double velAngle = std::atan2(circ.circle.velocity.y, circ.circle.velocity.x);
        auto circShape = std::make_shared<CollisionsShapes::Circle>(
            Position(circ.circle.center.x, circ.circle.center.y, velAngle),
            circ.circle.radius
        );
        auto velocity = std::make_shared<ObstacleVelocity>(
            0.0,
            0.0,
            velAngle,
            velDist
        );
        newLidar.emplace_back(std::make_shared<Obstacle>(circShape, velocity));
    }
    if (!newLidar.empty()) {
        obstaclesStack_->updateLidarObjects(newLidar);
    }
}

Position CollisionsSubscriptions::updateRobotPos()
{
    double tx, ty, rz;
    try {
        auto transform = tf2PosBuffer_.lookupTransform(MAP_TF_FRAME, ROBOT_TF_FRAME, ros::Time());
        tx = transform.transform.translation.x;
        ty = transform.transform.translation.y;
        rz = radToDegrees(
            quaternionToEuler(transform.transform.rotation).getAngle()
        );
    } catch (tf2::TransformException &ex) {
        ROS_WARN_STREAM("Error when trying to get /map/robot tf: " << ex.what());
    } catch (...) {
        ROS_ERROR_ONCE("An unknown exception happened when trying to update robot's position from tf. This message will be printed once.");
        tx = ty = rz = 0.0;
    }
    return { tx, ty, rz };
}


std::string CollisionsSubscriptions::fetchRobotName(ros::NodeHandle& nodeHandle)
{
    std::string robot_name; // TODO use C++17 init in if statement
    if (nodeHandle.getParam("/robot", robot_name))
    {
        return robot_name;
    }
    ROS_WARN_STREAM("Error when trying to get '/robot' param. Falling back to default name \"" << DEFAULT_ROBOT_NAME << "\".");
    return DEFAULT_ROBOT_NAME;
}

Position quaternionToEuler(geometry_msgs::Quaternion quaternion)
{
    double qw, qx, qy, qz;
    qw = quaternion.w;
    qx = quaternion.x;
    qy = quaternion.y;
    qz = quaternion.z;
    
    double t0 = +2.0 * (qw * qx + qy * qz);
    double t1 = +1.0 - 2.0 * (qx * qx + qy * qy);
    double ex = std::atan2(t0, t1);
    
    double t2 = +2.0 * (qw * qy - qz * qx);
    if (t2 < -1.0)
        t2 = -1.0;
    else if (t2 > 1.0)
        t2 = 1.0;
    double ey = std::asin(t2);
    
    double t3 = +2.0 * (qw * qz + qx * qy);
    double t4 = +1.0 - 2.0 * (qy * qy + qz * qz);
    double ez = std::atan2(t3, t4);
    
    return {ex, ey, ez};
}
